src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -12,202 +12,390 @@
                 continue;
             }
 
-            let mut next_offset = idx + 1;
+            let mut compaction_bytes = level.total_file_size;
+            let mut select_level_inputs = vec![InputLevel {
+                level_idx: 0,
+                level_type: level.level_type,
+                table_infos: level.table_infos.clone(),
+            }];
             let max_compaction_bytes = std::cmp::min(
                 self.config.max_compaction_bytes,
                 self.config.max_bytes_for_level_base,
             );
 
-            for other in &levels[0].table_infos[idx + 1..] {
-                if level_handlers[0].is_pending_compact(&other.id) {
+            for other in &l0.sub_levels[idx + 1..] {
+                if compaction_bytes >= max_compaction_bytes {
                     break;
                 }
-                // no need to trigger a bigger compaction
-                if compaction_bytes >= max_compaction_bytes {
+
+                if level.total_file_size > self.config.sub_level_max_compaction_bytes {
                     break;
                 }
-                compaction_bytes += other.file_size;
-                select_level_inputs.push(other.clone());
-                next_offset += 1;
-            }
 
-            // to make compaction tree balance, we do not need to merge a large file and a few small
-            // files.
-            while let Some(first) = select_level_inputs.first() {
-                if first.file_size * 2 > compaction_bytes
-                    && select_level_inputs.len()
-                        >= self.config.level0_tier_compact_file_number as usize
-                    && compaction_bytes > MIN_COMPACTION_BYTES
+                if other
+                    .table_infos
+                    .iter()
+                    .any(|table| level_handler.is_pending_compact(&table.id))
                 {
-                    compaction_bytes -= first.file_size;
-                    select_level_inputs.remove(0);
-                } else {
                     break;
                 }
+
+                compaction_bytes += other.total_file_size;
+                select_level_inputs.push(InputLevel {
+                    level_idx: 0,
+                    level_type: other.level_type,
+                    table_infos: other.table_infos.clone(),
+                });
             }
 
-            if select_level_inputs.len() < self.config.level0_tier_compact_file_number as usize {
-                idx = next_offset;
+            if select_level_inputs
+                .iter()
+                .map(|level| level.table_infos.len())
+                .sum::<usize>()
+                < self.config.level0_tier_compact_file_number as usize
+            {
                 continue;
             }
 
-            level_handlers[0].add_pending_task(self.compact_task_id, &select_level_inputs);
+            if select_level_inputs.len() <= self.config.level0_tier_compact_file_number as usize / 2
```

Comment:
`level0_tier_compact_file_number` controls not only the number of sub-level files but also the number of sub-levels to trigger L0 compaction. Why do we need to care about the number of sub-levels? Specifically, why not trigger compaction when the `sub-levels files > level0_tier_compact_file_number` but the `number of sub-levels < level0_tier_compact_file_number/2`?

Reply:
For example, if there are only two sub-level and each of them are non-overlapping. It is not necessary to merge them together.  Because it only decrease one level, but it make the data in this level compact once again. To reduce write-amplification, I hope we can merge small level with a large number.